src/Microsoft.SqlTools.ServiceLayer/LanguageServices/InteractionMetrics.cs

```diff
@@ -0,0 +1,97 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.SqlTools.ServiceLayer.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer
+{
+    /// <summary>
+    /// A class to calculate the value for the metrics using the given bucket
+    /// </summary>
+    public class InteractionMetrics<T>
+    {
+        /// <summary>
+        /// Creates new instance given a bucket of metrics
+        /// </summary>
+        public InteractionMetrics(int[] metrics)
+        {
+            Validate.IsNotNull("metrics", metrics);
+            if(metrics.Length == 0)
+            {
+                throw new ArgumentOutOfRangeException("metrics");
+            }
+
+            Counters = new ConcurrentDictionary<string, T>();
+            Metrics = IsSorted(metrics) ? metrics : metrics.OrderBy( i => i).ToArray();
+        }
+
+        private ConcurrentDictionary<string, T> Counters { get; }
+
+        private object perfCountersLock = new object();
+
+        /// <summary>
+        /// The metrics bucket
+        /// </summary>
+        public int[] Metrics { get; private set; }
+
+        /// <summary>
+        /// Returns true if the given list is sorted
+        /// </summary>
+        private bool IsSorted(int[] metrics)
+        {
+            if (metrics.Length > 1)
+            {
+                int previous = metrics[0];
+                for (int i = 1; i < metrics.Length; i++)
+                {
+                    if(metrics[i] < previous)
+                    {
+                        return false;
+                    }
+                    previous = metrics[i];
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Update metric value given new number
+        /// </summary>
+        public void UpdateMetrics(double duration, T newValue, Func<string, T, T> updateValueFactory)
+        {
+           Task.Factory.StartNew(() =>
```

Comment:
Counters is ConcurrentDictionary so it's thread safe and debugging was easier this way because without it the completion list would time out most of the times  

---
In reply to: [91594323](https://github.com/Microsoft/sqltoolsservice/pull/167#discussion_r91594323) [](ancestors = 91594323)